linera-core/src/worker.rs

```diff
@@ -398,18 +395,32 @@ where
         &self,
         certificate: LiteCertificate<'_>,
     ) -> Result<Either<ConfirmedBlockCertificate, ValidatedBlockCertificate>, WorkerError> {
-        match self
-            .recent_confirmed_value_cache
-            .full_certificate(certificate.clone())
+        if let Some(executed_block) = self
+            .recent_executed_block_cache
+            .get(&certificate.value.value_hash)
             .await
```

Comment:
You could add `.ok_or(WorkerError::MissingCertificateValue)?` here; saves some indentation and the reader doesn't have to remember at the end of the function what the `else` refers to.